### DISCUSSION TITLE
Best way to get properties of all physical elements

### DISCUSSION THREAD

**Discussion by lijiayi0921** [iTwin/itwinjs-core#3343] [Q&A]
Hi,

I'm trying to get a selection (by name for example) of attribute properties of all PhysicalElements present in a model.

For properties, I currently do so with presentation.getConent :

```
    // having acquired classFullName and id of all physical elements inside "props" :

    const options = {
        imodel, rulesetOrId: DEFAULT_PROPERTY_GRID_RULESET, keys: new KeySet(props.map(prop => {
            return {
                classFullName: prop.classFullName,
                id: prop.id,
            };
        }))
    };
    const descriptor = await Presentation.presentation.getContentDescriptor({ ...options, displayType: "Grid" });
    if (descriptor) {
        return await Presentation.presentation.getContent({ ...options, descriptor });
    }
```

This solution works. However the response json may be very large and takes a lot of time to download.

What would be the right way to get attribute properties names / values ?

Thanks
Jiayi

**Comment by grigasp** [comment]
You can specify what properties you want using the [fieldsSelector](https://github.com/iTwin/itwinjs-core/blob/dbab3d1837bc170fdcd2b72330288400341b8c6e/presentation/common/src/presentation-common/content/Descriptor.ts#L187) attribute when requesting content.

When requesting content for many elements I'd also recommend doing that in pages. We found that the most efficient way to do it is by paging the `KeySet` you're passing to `getContent`. We have a function on the backend that you could use as an example - [PresentationManager.getElementProperties](https://github.com/iTwin/itwinjs-core/blob/dbab3d1837bc170fdcd2b72330288400341b8c6e/presentation/backend/src/presentation-backend/PresentationManager.ts#L669-L699). 

I would also like to raise a question why you need to download properties of all `PhysicalElement` elements to the frontend. For large iModels that can be millions of elements and will never be efficient.

**Reply by lijiayi0921** [reply]
Thanks @grigasp for your answer !

I created a list of NamedFieldDescriptor by giving property field names : 

```
        let fields: NamedFieldDescriptor[] = [];
        for (const fn of fieldNames) {
            fields.push({
                type: FieldDescriptorType.Name,
                fieldName: fn,
            });
        }

        descOvw.fieldsSelector = {
            type: 'include',
            fields: fields,
        }

        await Presentation.presentation.getContent({ ...options, descriptor: descOvw }),
```

Appearantly when I put multiple fieldNames, getContent returns an empty list.
Am I using NamedFieldDescriptor correctly ?

----------------------------------------------------

What I try to achieve, is to port an application from Forge to iTwin, which takes some properties of all physical elements of a model when it is loaded, and computes the carbon footprint of it. The need is to rewrite the lower level API calls with itwinjs and reuse the same project configuration and computing logics.

Would you recommand other methods (presentation back-end, ecsql, ....) in my usecase ?

Thanks for your help
Jiayi

**Reply by ColinKerr** [reply]
The Presentation API is optimized to drive UI so it is good at getting the properties for single element (the property grid) or for many elements in a paged table.  It is not great for iterating over the values many elements because it the JSON has all the data needed for presentation.

Executing an ECSql query or queries for exactly the data you are looking for is the most efficient way to get the data on the frontend.  I'm not sure how familiar you are with ECSql so here is some info that can help you get started.  I gave two webinars about querying data in an iModel.  
 - See the first two videos in this playlist: https://www.youtube.com/playlist?list=PL6YCKeNfXXd9VFzTF-NSeN_l5vB7OTWR6. There is a transcript in the description of each talk.  
 - You can also find a lot of good information here: https://www.itwinjs.org/learning/ecsql/. 
 - Here are some examples of the query API: https://www.itwinjs.org/learning/frontend/ecsqlcodeexamples/

From your original description it sounds like you know the classes and maybe the properties that have the data you are looking for.  If so ECSql queries should be pretty straight forwards.  

If you do not then you would first need to find the properties and classes to search for.  You can do this with metadata queries like this one which will find all properties that have 'volume' in their name:
```sql
SELECT ec_classname(p.Class.Id) as classname, p.name FROM ECDbMeta.ECPropertyDef p WHERE p.Name LIKE '%volume%'
```

This solution is more complex than requesting all properties of every element but it will scale to much larger models than requesting all properties of each element.  If you need more help please let us know.

**Reply by grigasp** [reply]
> Thanks @grigasp for your answer !
> 
> I created a list of NamedFieldDescriptor by giving property field names :
> 
> ```
>         let fields: NamedFieldDescriptor[] = [];
>         for (const fn of fieldNames) {
>             fields.push({
>                 type: FieldDescriptorType.Name,
>                 fieldName: fn,
>             });
>         }
> 
>         descOvw.fieldsSelector = {
>             type: 'include',
>             fields: fields,
>         }
> 
>         await Presentation.presentation.getContent({ ...options, descriptor: descOvw }),
> ```
> 
> Appearantly when I put multiple fieldNames, getContent returns an empty list. Am I using NamedFieldDescriptor correctly ?
> 
> What I try to achieve, is to port an application from Forge to iTwin, which takes some properties of all physical elements of a model when it is loaded, and computes the carbon footprint of it. The need is to rewrite the lower level API calls with itwinjs and reuse the same project configuration and computing logics.
> 
> Would you recommand other methods (presentation back-end, ecsql, ....) in my usecase ?
> 
> Thanks for your help Jiayi

Can you please create a [sandbox](https://www.itwinjs.org/sandboxes/) to demonstrate the issue?

**Reply by lijiayi0921** [reply]
@grigasp I created this sandbox which should demonstrate the issue, but it has some errors...
https://www.itwinjs.org/sandboxes/Jiayi/Jillayne-the-aqua-owl-882

**Reply by lijiayi0921** [reply]
Thanks @ColinKerr for the answer. I'll let you know when I have more questions :)

**Reply by grigasp** [reply]
> @grigasp I created this sandbox which should demonstrate the issue, but it has some errors... https://www.itwinjs.org/sandboxes/Jiayi/Jillayne-the-aqua-owl-882

@lijiayi0921, please have a look at this: https://www.itwinjs.org/sandboxes/grigas/Opaline-the-gray-kite-479

**Reply by lijiayi0921** [reply]
Thanks a alot @grigasp !
I rewrote elemProps with ecsql in order to improve the performance.
Code creating an array of element items is added to log all items with properties.

In the model "CoffsHarborDemo", I look for 2 fields : 'Sloped Area', 'Name'
The result array does not contain those properties except for the first item

**Reply by grigasp** [reply]
Let's break that down by property:

- `Sloped Area` property exists only on the `orddgn.VolumetricQuantityAspect` class. I see there are some of its instances, but none of them have a relationship to a `bis.PhysicalElement` instance. This is the reason you're not even getting this property in the content descriptor.

- `Name` property comes from the relationship path `bis.Element -> bis.ModelContainsElements -> bis.Model -> bis.ModelModelsElement -> bis.PhysicalPartition -> bis.ElementHasLinks -> bis.RepositoryLink.UserLabel`. An content item is only going to have this property fs the target instance exists on this relationship path. I ran this query:

  ```sql
  select count(1)
  from bis.physicalelement pe
  join bis.physicalpartition pp on pp.ecinstanceid = pe.model.id
  join bis.ElementHasLinks ehl on ehl.sourceecinstanceid = pp.ecinstanceid
  join bis.repositorylink rl on rl.ecinstanceid = ehl.targetecinstanceid
  ```
   It tells me there are 3 physical elements whose model has a related repository link. I looked at the stuff you log into console and confirmed there are 3 records that have this property. Indexes in the array: 0, 677, 678.

So as far as I can tell everything works as expected.

**Reply by lijiayi0921** [reply]
@grigasp Thanks now I understand better how it works.
I wrote a comment with examples of fields causing returned contentSet to be empty. But github keeps saying "You can't comment at this time."
It happends when I test with an ifc file
How can I send you the example ?
Thanks
